cmd/build.go

```diff
@@ -0,0 +1,125 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/boson-project/faas"
+	"github.com/boson-project/faas/buildpacks"
+	"github.com/boson-project/faas/docker"
+	"github.com/ory/viper"
+	"github.com/spf13/cobra"
+)
+
+var buildCmd = &cobra.Command{
+	Use:        "build",
+	Short:      "Build an existing function project as an OCI image",
+	SuggestFor: []string{"biuld", "buidl", "built", "image"},
+	// TODO: Add completions for build
+	// ValidArgsFunction: CompleteRuntimeList,
+	RunE: buildImage,
+	PreRunE: func(cmd *cobra.Command, args []string) (err error) {
+		flags := []string{"path", "tag", "push"}
+		for _, f := range flags {
+			err = viper.BindPFlag(f, cmd.Flags().Lookup(f))
+			if err != nil {
+				return err
+			}
+		}
+		return
+	},
+}
+
+func init() {
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	root.AddCommand(buildCmd)
+	buildCmd.Flags().StringP("path", "p", cwd, "Path to the function project directory")
+	buildCmd.Flags().StringP("tag", "t", "", "Specify an image tag, for example quay.io/myrepo/project.name:latest")
+}
+
+type buildConfig struct {
+	Verbose bool
+	Path    string
+	Tag     string
+	Push    bool
+}
+
+func buildImage(cmd *cobra.Command, args []string) (err error) {
+	var config = buildConfig{
+		Verbose: viper.GetBool("verbose"),
+		Path:    viper.GetString("path"),
+		Tag:     viper.GetString("tag"),
+	}
+	return Build(config)
+}
+
+// FunctionConfigForBuild accepts a buildConfig and
+// uses this to load a faas.Function. If the config contains
+// a value for the image tag, this will be used and
+// subsequently written to the config file.
+func FunctionConfigForBuild(config buildConfig) (f *faas.Function, err error) {
```

Comment:
There's some logic in here in `FunctionConfigForBuild` and `Build` that is not really CLI-specific
- If Name and Runtime are not provided, generate an error
- If a Tag is not provided, use the tag from the config file in the function root
- Tags default to quay.io/%s:latest
- Update the Function config file on disk at function root
- Optionally push on Build

These logic bits are simple, yes, but they are applicable to a developer using the client library directly or the CLI.  So perhaps this is a job for a future refactor, but it would be nice to see these moved into the Client Library, and the API used here in the CLI kept to a very simple "map in booleans and strings, then invoke Build".  This would both remove the need to do a conditional invocation of the client library (one with and one without a Pusher), as well as opens us up to writing tests that validate expected functionality such as:
"Test that Name and Runtime are required or error"
"Test that Tag defaults to DefaultTag (quay.io)"
"Test that a Builder respects the Push boolean configuration"
etc.

Reply:
Excellent points. I have moved it in to function.go and renamed it to `faas.FunctionConfiguration()`. I think it might benefit from adding a `name` parameter as well in order to update the function name. But to be honest, it's a little clumsy right now and I'm not sure if this is the best way to approach it all in the long run.

Reply:
We probably have several solid refactors ahead of us where we can make it less clumsy!